File: codigo/1D/EF_barra_n_nodos/comparing_interpolation_algorithms.m
% Interpolation algorithms already implemented in MATLAB
%
% WHO   DATE            WHAT
% DAA   Mar 1,  2010    First algorithm
% DAA   Nov 11, 2021    Introducing MAKIMA
%
% DAA - Diego Andres Alvarez Marin - daalvarez@unal.edu.co

% Input the data points
n = 10; %number of data points
x = zeros(n,1);
y = zeros(n,1);
figure
title(sprintf('Click with the mouse %d times. The interpolation will pass through those points',n))
axis([-5 5 -5 5]);
hold on;
for i = 1:n
   [x(i),y(i)] = ginput(1);
   plot(x(i), y(i), 'ko','MarkerSize',12);
   plot(x(i), y(i), 'kx','MarkerSize',12);
end

xi = -10:0.1:10;
yi = interp1(x,y,xi,'nearest','extrap'); h1 = plot(xi,yi,'b.', 'LineWidth',2);
yi = interp1(x,y,xi,'linear', 'extrap'); h2 = plot(xi,yi,'r-', 'LineWidth',2);
yi = interp1(x,y,xi,'spline', 'extrap'); h3 = plot(xi,yi,'m-', 'LineWidth',2);
yi = interp1(x,y,xi,'makima', 'extrap'); h4 = plot(xi,yi,'r--','LineWidth',2);
yi = interp1(x,y,xi,'pchip',  'extrap'); h5 = plot(xi,yi,'g-', 'LineWidth',2);
% The 'cubic' method does not support extrapolation.
yi = interp1(x,y,xi,'cubic'           ); h6 = plot(xi,yi,'c-', 'LineWidth',2);

% Polynomial curve fitting (LAGRANGE POLYNOMIAL)
% Calculate the coefficients in the approximating polynomial of degree n-1
p = polyfit(x,y,n-1);
h7 = plot(xi,polyval(p,xi),'k-','LineWidth',4);
axis([-10  10  -5  5]);

legend([h1 h2 h3 h4 h5 h6 h7], ...
    'nearest','linear','spline','makima','pchip','cubic','Lagrange', ...
    'Location','NorthEastOutside')

% TAREA: leer la ayuda del los comandos pchip, makima, spline, interp1

File: codigo/1D/EF_barra_n_nodos/comparing_interpolation_algorithms.py
# Interpolation algorithms already implemented in SCIPY
#
# WHO   DATE            WHAT
# DAA   Oct  6, 2020      First algorithm
# DAA   May 30, 2025      Added Akima and Makima
#
# DAA - Diego Andres Alvarez Marin - daalvarez@unal.edu.co

import numpy as np
import matplotlib.pyplot as plt
# %matplotlib qt5
from scipy.interpolate import (
    PchipInterpolator, 
    CubicSpline, 
    Akima1DInterpolator, 
    interp1d, 
    lagrange
)    
# This script shows how to use different interpolation algorithms
# to interpolate a set of points in 1D. 

# The user clicks on the plot to input the data points.
# The algorithms are:
# - PchipInterpolator         : Piecewise Cubic Hermite Interpolating Polynomial
# - CubicSpline               : Cubic spline interpolation
# - Akima1DInterpolator       : Akima's method for 1D interpolation
# - interp1d with 'nearest'   : Nearest neighbor interpolation
# - interp1d with 'linear'    : Linear interpolation
# - interp1d with 'quadratic' : Quadratic interpolation
# - interp1d with kind=5      : Polynomial interpolation of degree 5
# - lagrange                  : Lagrange polynomial interpolation


# Input the data points
n = 10           #number of data points
plt.figure()
plt.title(f'Click with the mouse {n} times. The interpolation will pass through those points')
plt.axis([-5, 5, -5, 5])
pts = np.asarray(plt.ginput(n, timeout=-1, mouse_stop=None))
x, y = pts[:,0], pts[:,1]
plt.close()

algorithms = [
    (PchipInterpolator(  x, y,                       extrapolate=True),         'Pchip'      ),
    (CubicSpline(        x, y, bc_type='not-a-knot', extrapolate=True),         'CubicSpline'),
    (Akima1DInterpolator(x, y, method='akima'),                                 'Akima'      ),
    (Akima1DInterpolator(x, y, method='makima'),                                'Makima'     ),
    (interp1d(           x, y, kind='nearest',       fill_value='extrapolate'), 'nearest'    ),
    (interp1d(           x, y, kind='linear',        fill_value='extrapolate'), 'linear'     ),
    (interp1d(           x, y, kind='quadratic',     fill_value='extrapolate'), 'quadratic'  ),
#   (interp1d(           x, y, kind='cubic',         fill_value='extrapolate'), 'cubic'      ), # same as CubicSpline
    (interp1d(           x, y, kind=5,               fill_value='extrapolate'), '5'          ),
    (lagrange(           x,y),                                                  'Lagrange'   ),
# There are many more functions, as documented in:
# https://docs.scipy.org/doc/scipy/reference/interpolate.html
]
    
plt.figure()
plt.plot(x, y, 'ro')
xi = np.linspace(-10, 10, 501)
for alg in algorithms:
    yi = alg[0](xi)
    if   alg[1] == 'nearest':
        plt.plot(xi, yi, '.',  linewidth=2, label=alg[1])
    elif alg[1] == 'Lagrange':
        plt.plot(xi, yi, 'k-', linewidth=4, label=alg[1])        
    else:
        plt.plot(xi, yi,       linewidth=2, label=alg[1])

plt.legend(loc='best')
plt.ylim([-5, 5])
plt.show()

# More examples here: https://mmas.github.io/interpolation-scipy
